src/passes/OptimizeInstructions.cpp

```diff
@@ -133,6 +133,69 @@ struct LocalScanner : PostWalker<LocalScanner> {
   }
 };
 
+// perform some final optimizations
+struct FinalOptimizer
+  : public WalkerPass<
+      PostWalker<FinalOptimizer, UnifiedExpressionVisitor<FinalOptimizer>>> {
+  const PassOptions& passOptions;
+
+  FinalOptimizer(const PassOptions& passOptions) : passOptions(passOptions) {}
+
+  void visitExpression(Expression* curr) {
+    // if this contains dead code, don't bother trying to optimize it, the type
+    // might change (if might not be unreachable if just one arm is, for
+    // example). this optimization pass focuses on actually executing code. the
+    // only exceptions are control flow changes
+    if (curr->is<Const>() || curr->is<Call>() || curr->is<Nop>() ||
+        (curr->type == Type::unreachable && !curr->is<Break>() &&
+         !curr->is<Switch>() && !curr->is<If>())) {
```

Comment:
I think all these checks are not needed. They can be removed, and this function merged with `optimize`, so there is just one function, `visitExpression`.